lib/publiccloud/provider.pm

```diff
@@ -390,10 +394,14 @@ sub terraform_apply {
     if (get_var('PUBLIC_CLOUD_SLES4SAP')) {
         $vms = $output->{cluster_nodes_name}->{value};
         $ips = $output->{cluster_nodes_public_ip}->{value};
-        my $iscsi_vms = $output->{iscsisrv_name}->{value};
-        my $iscsi_ips = $output->{iscsisrv_public_ip}->{value};
-        push @{$vms}, @{$iscsi_vms};
-        push @{$ips}, @{$iscsi_ips};
+        foreach my $others_vms (qw(drbd iscsisrv)) {
+            my $tmp_name_var = "${others_vms}_name";
+            my $tmp_ip_var   = "${others_vms}_public_ip";
+            my $tmp_vms      = $output->{${tmp_name_var}}->{value};
+            my $tmp_ips      = $output->{${tmp_ip_var}}->{value};
+            push @{$vms}, @{$tmp_vms} if @{$tmp_vms};
+            push @{$ips}, @{$tmp_ips} if @{$tmp_ips};
```

Comment:
The if condition in these lines may be required to avoid issues if the arrayref is not defined.

Safest workaround would be to use `if (defined $tmp_ips and ref $tmp_ips eq 'ARRAY')`, but if you're sure these variables contain an arrayref, I would suggest dropping the if-condition altogether.

Reply:
Current workaround will not push the new elements if the array is empty. I don't believe you want that.

Reply:
> Current workaround will not push the new elements if the array is empty. I don't believe you want that.

This is what I want: if variables are empty I don't want to push them, because it's not needed (and it causes error also). And I'm doing this because there is no DRBD code for AWS at that time, so no DRBD results (this will change in a future version).

Reply:
> This is what I want: if variables are empty I don't want to push them, because it's not needed (and it causes error also). And I'm doing this because there is no DRBD code for AWS at that time, so no DRBD results (this will change in a future version).

Then definitely go with `if @{$tmp_ips}`. No defined needed.

Reply:
That's the issue I had and that's why I added `if defined` (http://1b210.qa.suse.de/tests/6492#step/sles4sap/132): `# Test died: Can't use an undefined value as an ARRAY reference at /var/lib/openqa/cache/1b210.qa.suse.de/tests/sle/lib/publiccloud/provider.pm line 402.`

And i still have it, it seems that `if @array` is not exactly the same as `if defined @array`...

Reply:
It's not the same. `defined` checks whether the variable has been initialized/defined, while without `defined` it will attempt to perform a boolean operation on the variable. For example:

```
my $var;
print "Variable is defined\n" if defined $var;
print "Variable is true\n" if $var;
```

Will not print anything as the value of the variable is undef. In this case both `if defined $var` and `if $var` give the same result, but with a small change, for example:

```
my $var = 0;
print "Variable is defined\n" if defined $var;
print "Variable is true\n" if $var;
```

The message `Variable is defined` is shown. If instead, `$var` is assigned to a positive number or to a non-empty string, both `Variable is defined` and `Variable is true` will get printed.

In this case, I think the error comes not from the `push` part of the line, but from the `if @{$var}` part of the line.

My suggestion would be to play it safe:

```
            push @{$vms}, @{$tmp_vms} if defined $tmp_vms and ref $tmp_vms eq 'ARRAY';
            push @{$ips}, @{$tmp_ips} if defined $tmp_ips and ref $tmp_ips eq 'ARRAY';
```

And let's cross fingers and this doesn't violate any of the travis' tests rules.

Reply:
Or better yet, @cfconrad 's workaround from below. :)

Reply:
I think the problem is that you retrieve `undef` from `$output->{$tmp_name_var}->{value}` and then you get that error on `@{$tmp_vms}`
See above my way of solving it, if you don't care if they are not present.

Reply:
> In this case, I think the error comes not from the `push` part of the line, but from the `if @{$var}` part of the line.

The error was from the `push` I think, as I first had it without any `if` call.

Reply:
> The error was from the `push` I think, as I first had it without any `if` call.

Yes, without the if, it must come from the push. But with the if, I think it comes from the if part as it's evaluated first.

What happens is that the variable has a value of `undef`, and when you do `@{$var}` it's basically trying to access an array referenced by a variable that doesn't exist. So either with the `@{$var}` as an argument to the push, or as an argument to the if, error is the same: cannot use undefined variable as ARRAYREF.

Reply:
`if defined $tmp_vms` seems to be correct also. If all you you agree I will go with this.